ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -654,34 +677,37 @@ private void updateVersionStateAndDeleteDatasetVersionOutOfRetentionCount() thro
      * @param accountName the name of the account.
      * @param containerName the name of the container.
      * @param datasetName the name of the dataset.
+     * @param wrappedRestRequest the {@link WrappedRestRequest} which always return DELETE for getRestMethod.
      */
-    private Callback<Void> recursiveCallback(List<DatasetVersionRecord> datasetVersionRecordList, int idx, String accountName,
-        String containerName, String datasetName) {
+    private Callback<Void> recursiveCallback(List<DatasetVersionRecord> datasetVersionRecordList, int idx,
+        String accountName, String containerName, String datasetName, Callback<Void> processResponseCallback,
+        WrappedRestRequest wrappedRestRequest) {
       if (idx == datasetVersionRecordList.size()) {
         LOGGER.debug("Complete recursive callback for " + idx + " number of dataset version");
         return (r, e) -> {
-          //In the last callback, set the rest method back.
-          LOGGER.debug("Change rest method back to PUT");
-          restRequest.setRestMethod(RestMethod.PUT);
+          //make sure to process response delete is done.
+          processResponseCallback.onCompletion(null, e);
         };
       }
       Callback<Void> nextCallBack =
-          recursiveCallback(datasetVersionRecordList, idx + 1, accountName, containerName, datasetName);
+          recursiveCallback(datasetVersionRecordList, idx + 1, accountName, containerName, datasetName,
+              processResponseCallback, null);
```

Comment:
wrapped request shouldn't be null in this function call.

Reply:
Thanks for catching that, fixed it and the updated test should cover this case.